nomad/structs/structs.go

```diff
@@ -11428,6 +11438,84 @@ func (a *Allocation) NeedsToReconnect() bool {
 	return disconnected
 }
 
+// LeaderAndMainTasksInGroup returns two slices: one with the leader tasks and
+// another with the main tasks that are not leader in the given task group.
+// If the task group is no longer present in the allocation definition
+// or there are no leaders and the main body the task is empty both slices will
+// be nil. If the task group has only one task, both slices will contain that task.
+//
+// This function is optimized to avoid traversing the task group tasks more than
+// once in case there is no leader defined.
+func (a *Allocation) LeaderAndMainTasksInGroup(tg *TaskGroup) ([]*Task, []*Task) {
+	if tg == nil {
+		return nil, nil
+	}
+
+	switch len(tg.Tasks) {
+	case 0:
+		return nil, nil
+
+	case 1:
+		return tg.Tasks, tg.Tasks
+	}
+
+	var leaderTasks, mainTasks []*Task
+
+	for _, t := range tg.Tasks {
+		if t.Leader {
+			leaderTasks = append(leaderTasks, t)
+			continue
+		}
+
+		if t.IsMain() {
+			mainTasks = append(mainTasks, t)
+		}
+	}
+
+	return leaderTasks, mainTasks
+}
```

Comment:
Given that this method is only used in the reconciler (a part of our code that is called _very_ often), it may be better to avoid allocating the slices and have the loop in `StartOfOldestTask`. Then we only need to keep the oldest time stored. 

Some untested sample code:

```go
func (a *Allocation) StartOfOldestTask(tg *TaskGroup) time.Time {
	now := time.Now().UTC()
	oldestStart := now

	for _, t := range tg.Tasks {
		if t.Leader {
			return a.LastStartOfTask(t.Name)
		}

		if t.IsMain() {
			ls := a.LastStartOfTask(t.Name)
			if !ls.IsZero() && ls.Before(oldestStart) {
				oldestStart = ls
			}
		}
	}

	return oldestStart
}
```

Since a group can only have a one leader, this also has the added benefit of cutting the loop short as soon as a leader is found.
https://github.com/hashicorp/nomad/blob/428103ba1265f519955d92817cdebeb673b1e58b/nomad/structs/structs.go#L6984-L6986